File: copyListrandom.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if head is None: return None
        current = head

        # connecting the list
        while current is not None:
            deepcopyNode = Node(current.val)
            deepcopyNode.next = current.next
            current.next = deepcopyNode
            current = current.next.next

        # connecting the Random Pointer of deep

        current = head
        while current is not None:
            if current.random is not None:
                current.next.random = current.random.next
            current = current.next.next

        # split the list
        current = head
        random = head.next
        randomptr = head.next

        while current is not None:
            current.next = current.next.next
            if current.next is not None:
                randomptr.next = randomptr.next.next

            current = current.next
            randomptr = randomptr.next

        return random

File: copylistrandom.java
/*
TC O(N)
SC O(1)
*/
class Solution {
    public Node copyRandomList(Node head) {
		
        Map<Node, Node> originalCopyMap = new HashMap<>();
        originalCopyMap.put(null, null);
        Node copyHead = new Node(10001);
        Node temp = copyHead;
        while(head != null) {
			
            Node copyNode = originalCopyMap.getOrDefault(head, null);
            if(copyNode == null) {
                copyNode = new Node(head.val);
                originalCopyMap.put(head, copyNode);
            }
            temp.next = copyNode;
            temp = temp.next;
            
            if(head.random != null) {
                Node copyRandom = originalCopyMap.getOrDefault(head.random, null);
                if(copyRandom == null) {
                    copyRandom = new Node(head.random.val);
                    originalCopyMap.put(head.random, copyRandom);
                }
                temp.random = copyRandom;
            }
            head = head.next;
        }
        return copyHead.next;  
    }
}